main/init/maven/src/mill/main/maven/MavenBuildGenMain.scala

```diff
@@ -0,0 +1,272 @@
+import mill.main.buildgen.*
+import mill.main.buildgen.BuildGenUtil.*
+import org.apache.maven.model.{Dependency, Model, Parent}
+import os.{Path, SubPath}
+
+import scala.jdk.CollectionConverters.*
+
+/**
+ * Converts a Maven build to Mill by generating Mill build file(s) from POM file(s).
+ *
+ * The generated output should be considered scaffolding and will likely require edits to complete conversion.
+ *
+ * ===Capabilities===
+ * The conversion
+ *  - handles deeply nested modules
+ *  - captures project settings
+ *  - configures dependencies for scopes:
+ *    - compile
+ *    - provided
+ *    - runtime
+ *    - test
+ *  - configures testing frameworks:
+ *    - JUnit 4
+ *    - JUnit 5
+ *    - TestNG
+ *  - configures multiple, compile and test, resource directories
+ *
+ * ===Limitations===
+ * The conversion does not support:
+ *  - plugins, other than maven-compiler-plugin
+ *  - packaging, other than jar, pom
+ *  - build extensions
+ *  - build profiles
+ */
+@mill.api.internal
+object MavenBuildGenMain extends BuildGenBase[Model, Dependency] {
+  type C = MavenBuildGenMain.Config
+
+  def main(args: Array[String]): Unit = {
+    val cfg = ParserForClass[Config].constructOrExit(args.toSeq)
+    run(cfg)
+  }
+
+  private def run(cfg: Config): Unit = {
+    val workspace = os.pwd
+
+    println("converting Maven build")
+    val modeler = Modeler(cfg)
+    val input = Tree.from(Seq.empty[String]) { dirs =>
+      val model = modeler(workspace / dirs)
+      (Node(dirs, model), model.getModules.iterator().asScala.map(dirs :+ _))
+    }
+
+    convertWriteOut(cfg, cfg.shared, input)
+
+    println("converted Maven build to Mill")
+  }
+
+  def getBaseInfo(
+      input: Tree[Node[Model]],
+      cfg: Config,
+      baseModule: String,
+      packagesSize: Int
+  ): IrBaseInfo = {
+    val model = input.node.value
+    val javacOptions = Plugins.MavenCompilerPlugin.javacOptions(model)
+    val pomSettings = extractPomSettings(model)
+    val publishVersion = model.getVersion
+    val publishProperties = getPublishProperties(model, cfg)
+
+    val typedef = IrTrait(
+      cfg.shared.jvmId,
+      baseModule,
+      getModuleSupertypes(cfg),
+      javacOptions,
+      pomSettings,
+      publishVersion,
+      publishProperties,
+      repos = Nil
+    )
+
+    // Model.getRepositories is always empty
+    IrBaseInfo(javacOptions, Seq.empty, noPom = false, publishVersion, publishProperties, typedef)
+  }
+
+  override def extractIrBuild(
+      cfg: Config,
+      baseInfo: IrBaseInfo,
+      build: Node[Model],
+      packages: Map[(String, String, String), String]
+  ): IrBuild = {
+    val project = build.value
+    val scopedDeps = extractScopedDeps(project, packages, cfg)
+    val version = project.getVersion
+    IrBuild(
+      scopedDeps = scopedDeps,
+      testModule = cfg.shared.testModule,
+      hasTest = os.exists(getMillSourcePath(project) / "src/test"),
+      dirs = build.dirs,
+      repositories = Nil,
```

Comment:
@ajaychandran maybe we can leave a comment here as to why `repositories = Nil` in the Maven build config? Is it because Maven doesn't support repositories? Or is it just an implementation limitation?

Reply:
I missed the change here in am earlier attempt. Support has been added.